### PR TITLE
Correctly handle wares carried into harbors with no path to goal

### PR DESCRIPTION
**Pull request by Flamefire**
When the goal for a ware becomes unreachable while it is carried into a harbor
`RecalcRoute` will call `FindRouteToWarehouse`.
Usually the selected warehouse will be the harbor as it is trivially the closest.
However when the harbor does not accept this ware type another one might be chosen
and the ware needs to be moved out instead of stored.

But `FindRouteToWarehouse` does NOT set `next_dir` while the ware is carried even when
called from `RecalcRoute` where the caller assumes it is set.
Changing this causes replays to go async, so work around it by calling `RecalcRoute` again
which now succeeds as it reached the goal in the previous call.

This bug also affects `WantInBuilding` which also calls `RecalcRoute` assuming it does so in all cases.

Fixes https://github.com/Return-To-The-Roots/s25client/issues/1843

The test case was a bit tricky to write and I added the case where no goal is possible in which case the ware should be accepted in the ware house anyway.

### CODE REVIEW

**Comment by stefson** [comment]
it seems as if I drove the logistics into an interesting corner case? :-)

File: libs/s25main/figures/nofWarehouseWorker.cpp [hunk 1]
```diff
@@ -62,84 +60,78 @@ void nofWarehouseWorker::Draw(DrawPoint drawPt)
...
     if(!shouldBringWareIn)
     {
-        // Ware an der Fahne ablegen ( wenn noch genug Platz ist, 8 max pro Flagge!)
-        // außerdem ggf. Waren wieder mit reinnehmen, deren Zi­el zerstört wurde
-        // ( dann ist goal = location )
-        if(world->GetSpecObj<noFlag>(pos)->HasSpaceForWare() && carried_ware->GetGoal() != carried_ware->GetLocation()
-           && carried_ware->GetGoal() != wh)
+        // Put ware down at flag if enough space.
+        // Might need to take it back in if goal was destroyed or changed to the warehouse
+        if(flag->HasSpaceForWare() && carried_ware->GetGoal() && carried_ware->GetGoal() != wh)
```
Comment (Flow86) [inline libs/s25main/figures/nofWarehouseWorker.cpp:70]: can there be no flag? (e.g `flag` being null?)
we should at least assert it

File: libs/s25main/figures/nofWarehouseWorker.cpp [hunk 2]
```diff
@@ -62,84 +60,78 @@ void nofWarehouseWorker::Draw(DrawPoint drawPt)
...
         {
-            carried_ware->WaitAtFlag(world->GetSpecObj<noFlag>(pos));
+            // TODO: Remove assert. Was added to verify prior condition
```
Comment (Flow86) [inline libs/s25main/figures/nofWarehouseWorker.cpp:72]: how do we resolve the TODO?

File: libs/s25main/Ware.cpp [hunk 3]
```diff
@@ -112,6 +112,7 @@ void Ware::RecalcRoute()
             static_cast<nobHarborBuilding*>(goal)->WareDontWantToTravelByShip(this);
         } else
         {
+            // TODO(Replay) This should calculate the next dir even when carried
```
Comment (Flow86) [inline libs/s25main/Ware.cpp:115]: theres another `TODO`?

File: libs/s25main/figures/nofWarehouseWorker.cpp [hunk 4]
```diff
@@ -62,84 +60,78 @@ void nofWarehouseWorker::Draw(DrawPoint drawPt)
...
+            // Add to warehouse inventory (if it still exists and was not burnt down)
             if(world->GetNO(pos)->GetType() == NodalObjectType::Building)
             {
                 auto* wh = world->GetSpecObj<nobBaseWarehouse>(pos);
-                if(carried_ware->GetGoal() == carried_ware->GetLocation() || carried_ware->GetGoal() == wh)
+                // Store the ware if its goal is this warehouse or it has no goal (anymore)
+                // Else it wants to go somewhere else, so add to waiting wares
+                if(!carried_ware->GetGoal() || carried_ware->GetGoal() == wh)
+                {
+                    // TODO: Remove assert. Was added to verify prior condition
```
Comment (Flow86) [inline libs/s25main/figures/nofWarehouseWorker.cpp:113]: another TODO?

**Review by Flow86 — CHANGES_REQUESTED** [review]

File: libs/s25main/figures/nofWarehouseWorker.cpp (hunk 1 above)
Comment (Flamefire) [reply libs/s25main/figures/nofWarehouseWorker.cpp:70]: If there was no flag, there is no warehouse and the worked would have been notified to abort the work. So yes there is always a flag

File: libs/s25main/figures/nofWarehouseWorker.cpp (hunk 2 above)
Comment (Flamefire) [reply libs/s25main/figures/nofWarehouseWorker.cpp:72]: Keep it for a short while, then remove. Possible to remove now as the auto-play tests succeeded but might be worth collecting a bit more evidence although I'm very sure this is correct.

File: libs/s25main/figures/nofWarehouseWorker.cpp (hunk 4 above)
Comment (Flamefire) [reply libs/s25main/figures/nofWarehouseWorker.cpp:113]: Same as above

File: libs/s25main/Ware.cpp (hunk 3 above)
Comment (Flamefire) [reply libs/s25main/Ware.cpp:115]: Yes: "TODO(Replay)" marks places that should be addressed when the replay version is increased, i.e. breaking old replays. I added this because it would be correct to do but can't be done as replays will go async

File: libs/s25main/figures/nofWarehouseWorker.cpp (hunk 1 above)
Comment (Flow86) [reply libs/s25main/figures/nofWarehouseWorker.cpp:70]: well, then we should assert it. I don't like non checked pointers

Comment (Flamefire) [reply libs/s25main/figures/nofWarehouseWorker.cpp:70]: Ok done. Also using `wh` where it makes sense to avoid fetching it multiple times and `wh->GetFlag()` is cleaner

**Review by Flow86 — APPROVED** [review]

**Comment by Flamefire** [comment]
> it seems as if I drove the logistics into an interesting corner case? :-)

Yeah, it's a timing issue and hence difficult to reproduce in "normal" circumstances so I cheated a bit for the test case.
